### PR TITLE
Upgrade to Lucene r1660560

### PR DESCRIPTION
**Pull request by rmuir**
This isnt an easy one. I deferred a lot of work until later.

There are a lot of followups we can do. I don't plan on working on these, or opening issues for these, they are just ideas.
1. TODO: needsScores: we aren't properly taking advantage of this, and something might not be right here
2. TODO: topdocs.merge, we are stricter and some loose stuff is happening.
3. Generate @Overrides for any overriden methods. This is a very useful tool when dealing with API changes/refactoring. It is crazy not to use it.
4. remove deprecated use of TermFilter
5. review "take 1" commits, we found a few unrelated bugs (e.g. assuming DIS.iterator() could not be null and other silly stuff)

One test is especially bad. It is not debuggable and I disabled it. It tries to use the postings lists (in wierd ways) from groovy apis or something like that. Besides the point of testing, this alone seems like a bad idea. We should remove tests like this from the codebase, they just waste time.  This test really ruined my day and discouraged me from working any further. Please consider the person debugging when writing tests.

### CODE REVIEW

**Comment by rjernst** [comment]
FYI, I re-enabled the non debuggable test (indeed it was very difficult to debug) and commented out the portions that cause problems, along with adding comments as to why they fail.

File: src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/InternalTopHits.java [hunk 1]
```diff
@@ -90,42 +90,86 @@ public SearchHits getHits() {
...
+
+        TopDocs reducedTopDocs;
         if (topDocs instanceof TopFieldDocs) {
-            sort = new Sort(((TopFieldDocs) topDocs).fields);
+            Sort sort = new Sort(((TopFieldDocs) topDocs).fields);
+
+            TopFieldDocs[] shardDocs = new TopFieldDocs[aggregations.size()];
+            for (int i = 0; i < shardDocs.length; i++) {
+                InternalTopHits topHitsAgg = (InternalTopHits) aggregations.get(i);
+                // TODO: topdocs.merge is this really safe?  our topDocs is a TopFieldDocs...
```
Comment (mikemccand) [inline src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/InternalTopHits.java]: @jpountz could you have a look at this one?  It made me nervous (not sure the stronger typing is safe).

File: src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/InternalTopHits.java [hunk 2]
```diff
@@ -90,42 +90,86 @@ public SearchHits getHits() {
         return searchHits;
     }
 
+    private TopFieldDocs asTopFieldDocs(TopDocs topDocs) {
+        if (topDocs instanceof TopFieldDocs) {
+            return (TopFieldDocs) topDocs;
+        } else {
+            // TODO: fix whoever is doing this!
```
Comment (mikemccand) [inline src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/InternalTopHits.java]: @jpountz maybe also this one?

**Comment by rmuir** [comment]
whew, awesome ryan! Thanks for all your help here (mike too). Another thing to mention, the "TODO: topdocs.merge" (there are two of these), we need to look into them. Mike had them as nocommits and maybe they should be...

**Comment by jpountz** [comment]
I'm looking into those TopDocs.merge issues.

**Comment by jpountz** [comment]
@mikemccand I pushed a new commit that hopefully makes usage of TopDocs.merge better. Let me know what you think?

**Comment by mikemccand** [comment]
Thanks @jpountz that looks much cleaner.

**Pull request closed without merging**